generator/src/main/java/org/apache/kafka/message/VersionConditional.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.message;
+
+/**
+ * Creates an if statement based on whether or not the current version
+ * falls within a given range.
+ */
+public final class VersionConditional {
+    /**
+     * Create a version conditional.
+     *
+     * @param containingVersions    The versions for which the conditional is true.
+     * @param possibleVersions      The range of possible versions.
+     * @return                      The version conditional.
+     */
+    static VersionConditional forVersions(Versions containingVersions,
+                                          Versions possibleVersions) {
+        return new VersionConditional(containingVersions, possibleVersions);
+    }
+
+    private final Versions containingVersions;
+    private final Versions possibleVersions;
+    private ClauseGenerator ifMember = null;
+    private ClauseGenerator ifNotMember = null;
+    private boolean alwaysEmitBlockScope = false;
+    private boolean allowMembershipCheckAlwaysFalse = true;
+
+    private VersionConditional(Versions containingVersions, Versions possibleVersions) {
+        this.containingVersions = containingVersions;
+        this.possibleVersions = possibleVersions;
+    }
+
+    VersionConditional ifMember(ClauseGenerator ifMember) {
+        this.ifMember = ifMember;
+        return this;
+    }
+
+    VersionConditional ifNotMember(ClauseGenerator ifNotMember) {
+        this.ifNotMember = ifNotMember;
+        return this;
+    }
+
+    VersionConditional alwaysEmitBlockScope(boolean alwaysEmitBlockScope) {
+        this.alwaysEmitBlockScope = alwaysEmitBlockScope;
+        return this;
+    }
+
+    VersionConditional allowMembershipCheckAlwaysFalse(boolean allowMembershipCheckAlwaysFalse) {
+        this.allowMembershipCheckAlwaysFalse = allowMembershipCheckAlwaysFalse;
+        return this;
+    }
```

Comment:
It might seem like a double negative, but it's actually not... it's just saying whether we should allow cases where the if statement is never taken.  There are a few conditionals where this should never happen, and this is just an easy way to set up that assertion (basically).  I will add some JavaDoc.

Reply:
I added more JavaDoc.